testnet/stacks-node/src/run_loop/neon.rs

```diff
@@ -394,156 +472,229 @@ impl RunLoop {
+            let ibd = match self.get_pox_watchdog().pox_sync_wait(
                 &burnchain_config,
                 &burnchain_tip,
-                if learned_burnchain_height {
-                    Some(burnchain_height)
-                } else {
-                    None
-                },
+                Some(remote_chain_height),
                 num_sortitions_in_last_cycle,
             ) {
                 Ok(ibd) => ibd,
                 Err(e) => {
-                    debug!("Pox sync wait routine aborted: {:?}", e);
+                    debug!("PoX sync wait routine aborted: {:?}", e);
                     continue;
                 }
             };
-            // will recalculate this
-            num_sortitions_in_last_cycle = 0;
 
-            let (next_burnchain_tip, next_burnchain_height) =
-                match burnchain.sync(Some(target_burnchain_block_height)) {
-                    Ok(x) => x,
-                    Err(e) => {
-                        warn!("Burnchain controller stopped: {}", e);
-                        continue;
-                    }
-                };
+            // will recalculate this in the following loop
+            num_sortitions_in_last_cycle = 0;
 
-            target_burnchain_block_height = cmp::min(
-                next_burnchain_height,
-                target_burnchain_block_height + pox_constants.reward_cycle_length as u64,
-            );
+            // Download each burnchain block and process their sortitions.  This, in turn, will
+            // cause the node's p2p and relayer threads to go fetch and download Stacks blocks and
+            // process them.  This loop runs for one reward cycle, so that the next pass of the
+            // runloop will cause the PoX sync watchdog to wait until it believes that the node has
+            // obtained all the Stacks blocks it can.
+            while burnchain_height <= target_burnchain_block_height {
+                if !self.should_keep_running.load(Ordering::SeqCst) {
+                    break;
+                }
 
-            // *now* we know the burnchain height
-            learned_burnchain_height = true;
-            burnchain_tip = next_burnchain_tip;
-            burnchain_height = next_burnchain_height;
+                let (next_burnchain_tip, tip_burnchain_height) =
+                    match burnchain.sync(Some(burnchain_height + 1)) {
+                        Ok(x) => x,
+                        Err(e) => {
+                            warn!("Burnchain controller stopped: {}", e);
+                            continue;
+                        }
+                    };
 
-            let sortition_tip = &burnchain_tip.block_snapshot.sortition_id;
-            let next_height = burnchain_tip.block_snapshot.block_height;
+                // *now* we know the burnchain height
+                burnchain_tip = next_burnchain_tip;
+                burnchain_height = cmp::min(burnchain_height + 1, tip_burnchain_height);
```

Comment:
What does limiting the increase of `burnchain_height` by 1 do in this loop? It seems like `burnchain_height` is only used as a parameter to `bitcoin_controller.sync(...)`, so each trip through the loop will always invoke with +1 of the prior invocation. If `burnchain.sync` returns a more advanced chain tip, why not progress burnchain_height to the new sortition height?

Reply:
This change makes sure that the burnchain indexer only downloads _one_ burnchain block per call.  The `.sync()` method takes an optional maximum burnchain block height, at which `.sync()` will return if it is reached.  Before, the node would download a whole reward cycle of burnchain blocks, and then run sortitions on all of them, which prevented the node from downloading Stacks blocks while burnchain blocks were being downloaded.  With this new loop, a node will download a burnchain block, process its sortition (waking up the chains coordinator), and then proceed to download the next burnchain block in the next loop pass while the p2p and relayer threads work on downloading the sortition's Stacks block.